Source/Core/Core/IOS/ES/NandUtils.cpp

```diff
@@ -200,65 +200,96 @@ std::vector<std::array<u8, 20>> ES::GetSharedContents() const
   return map.GetHashes();
 }
 
-bool ES::InitImport(u64 title_id)
+static bool DeleteDirectoriesIfEmpty(FS::FileSystem* fs, const std::string& path)
 {
-  const std::string content_dir = Common::GetTitleContentPath(title_id, Common::FROM_SESSION_ROOT);
-  const std::string data_dir = Common::GetTitleDataPath(title_id, Common::FROM_SESSION_ROOT);
-  for (const auto& dir : {content_dir, data_dir})
+  std::string::size_type position = std::string::npos;
+  do
   {
-    if (!File::IsDirectory(dir) && !File::CreateFullPath(dir) && !File::CreateDir(dir))
+    const auto directory = fs->ReadDirectory(0, 0, path.substr(0, position));
+    if ((directory && directory->empty()) ||
+        (!directory && directory.Error() != FS::ResultCode::NotFound))
     {
-      ERROR_LOG(IOS_ES, "InitImport: Failed to create title dirs for %016" PRIx64, title_id);
-      return false;
+      if (fs->Delete(0, 0, path.substr(0, position)) != FS::ResultCode::Success)
+        return false;
     }
+    position = path.find_last_of('/', position - 1);
+  } while (position != 0);
+  return true;
+}
+
+bool ES::InitImport(const IOS::ES::TMDReader& tmd)
+{
+  const auto fs = m_ios.GetFS();
+  const std::string content_dir = Common::GetTitleContentPath(tmd.GetTitleId());
+  const std::string import_content_dir = Common::GetImportTitlePath(tmd.GetTitleId()) + "/content";
+  if (fs->CreateFullPath(0, 0, content_dir + '/', 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite,
+                         FS::Mode::Read) != FS::ResultCode::Success ||
+      fs->SetMetadata(0, content_dir, 0, 0, 0, FS::Mode::ReadWrite, FS::Mode::ReadWrite,
+                      FS::Mode::None) != FS::ResultCode::Success ||
+      fs->CreateFullPath(0, 0, import_content_dir + '/', 0, FS::Mode::ReadWrite,
+                         FS::Mode::ReadWrite, FS::Mode::None) != FS::ResultCode::Success)
```

Comment:
This `if` kinda looks terrible...but I cannot really think of a nicer way to write this. Locals for the individual results won't really change much other than bloat those lines.

Reply:
Hmm, I agree. I've split the calls which I think looks a bit better (this is okay because we don't depend on short-circuit evaluation)